In [2]:
import pandas as pd
countries_df = pd.DataFrame({
    'country': ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    'population': [56.29, 38.05, 322.28, 146.24, 45.5, 9.5, 17.04],
    'square': [133396, 9984670, 9826630, 17125191, 603628, 207600, 2724902]
})

countries_df.to_csv('data/countries.csv', index=False, sep=';')

Экспорт данных в формат csv осуществляется с помощью метода DataFrame to_csv().  

Кликните на плашку, чтобы увидеть информацию ↓

Основные параметры метода DataFrame to_csv()

        path_or_buf — путь до файла, в который будет записан DataFrame (например, data/my_data.csv);
        sep — разделитель данных в выходном файле (по умолчанию ',');
        decimal — разделитель чисел на целую и дробную части в выходном файле (по умолчанию '.');

        columns — список столбцов, которые нужно записать в файл (по умолчанию записываются все столбцы);
        index — параметр, определяющий, требуется ли создавать дополнительный столбец с индексами строк в файле (по умолчанию True).



fff


In [3]:
countries_data = pd.read_csv('data/countries.csv', sep=';')
print(countries_data)

     country  population    square
0     Англия       56.29    133396
1     Канада       38.05   9984670
2        США      322.28   9826630
3     Россия      146.24  17125191
4    Украина       45.50    603628
5   Беларусь        9.50    207600
6  Казахстан       17.04   2724902


Запись и чтение в других форматах



Методы для записи таблиц в файлы отличных от csv форматов:

        to_excel() — запись DataFrame в формат Excel-таблицы (.xslx);
        to_json() — запись DataFrame в формат JSON (.json);
        to_xml() — запись DataFrame в формат XML-документа (.xml);
        to_sql() — запись DataFrame в базу данных SQL (для реализации этого метода необходимо установить соединение с базой данных).





Методы для чтения таблиц из файлов в отличных от csv форматах:

        read_excel() — чтение из формата Excel-таблицы(.xslx) в DataFrame;
        read_json() — чтение из формата JSON (.json) в DataFrame;
        read_xml() — чтение из формата XML-документа (.xml) в DataFrame;
        read_sql() — чтение из базы данных SQL в DataFrame (также необходимо установить соединение с базой данных).



Для этого у DataFrame есть методы head() и tail(), которые возвращают n первых и n последних строк таблицы соответственно (по умолчанию n = 5).

In [6]:
import pandas as pd
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
print(melb_data.head())


   index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...  1.0     202.0         126.0     1970.0   
1  4/02/2016       2.5  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5  ...  1.0      94.0         126.0     1970.0   
4  4/06/2016       2.5  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarr

In [13]:
print(melb_data.tail(7))

       index         Suburb        Address  Rooms Type      Price Method  \
13573  13573       Werribee    5 Nuragi Ct      4    h   635000.0      S   
13574  13574    Westmeadows     9 Black St      3    h   582000.0      S   
13575  13575  Wheelers Hill   12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown  77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown    83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown   96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville     6 Agnes St      4    h  1285000.0     SP   

             SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
13573  hockingstuart  26/08/2017      14.7  ...  1.0     662.0         172.0   
13574            Red  26/08/2017      16.5  ...  2.0     256.0         126.0   
13575          Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576       Williams  26/08/2017       6.8  ...  2.0     333.0         

Размерность таблицы

Далее хотелось бы узнать размер таблицы — количество строк и количество столбцов. Это можно сделать с помощью атрибута shape, который возвращает кортеж с количеством строк и столбцов:

In [14]:
print(melb_data.shape)

(13580, 23)


Получение информации о столбцах

Для того чтобы получить более детальную информацию о столбцах таблицы, можно использовать метод DataFrame info():




Данный метод выводит:

        информацию об индексах;
        информацию об общем количестве столбцов;
        таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Count) в каждом столбце и типе данных столбца (Dtype), количестве столбцов, в которых используется определённый тип данных;
        количество оперативной памяти в мегабайтах, которое тратится на хранение данных.

Из вывода метода info() становится понятно, что в нашей таблице есть столбец CouncilArea с пропущенными значениями — количество непустых значений в столбце меньше, чем количество строк в таблице (12211 < 13580).

Пустыми, или пропущенными, значениями называются значения в ячейках таблицы, которые не заполнены по какой-либо причине, то есть на их месте стоит пустое место. В Pandas такие значения обозначаются символом NaN (Not-a-Number).

→ Мы будем говорить о работе с пропусками в отдельном модуле, посвящённом очистке данных, а пока что не будем обращать на них внимания.

Если вам всё же любопытно, как можно работать с такими значениями в Pandas, то вы можете посмотреть информацию здесь.


In [5]:
print(melb_data.info)

<bound method DataFrame.info of        index         Suburb           Address  Rooms Type      Price Method  \
0          0     Abbotsford      85 Turner St      2    h  1480000.0      S   
1          1     Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2          2     Abbotsford      5 Charles St      3    h  1465000.0     SP   
3          3     Abbotsford  40 Federation La      3    h   850000.0     PI   
4          4     Abbotsford       55a Park St      4    h  1600000.0     VB   
...      ...            ...               ...    ...  ...        ...    ...   
13575  13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG    

In [8]:
melb_data['Car'] = melb_data['Car'].astype('int64')
melb_data['Bedroom'] = melb_data['Bedroom'].astype('int64')
melb_data['Bathroom'] = melb_data['Bathroom'].astype('int64')
melb_data['Propertycount'] = melb_data['Propertycount'].astype('int64')
melb_data['YearBuilt'] = melb_data['YearBuilt'].astype('int64')
print(melb_data.info)

<bound method DataFrame.info of        index         Suburb           Address  Rooms Type      Price Method  \
0          0     Abbotsford      85 Turner St      2    h  1480000.0      S   
1          1     Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2          2     Abbotsford      5 Charles St      3    h  1465000.0     SP   
3          3     Abbotsford  40 Federation La      3    h   850000.0     PI   
4          4     Abbotsford       55a Park St      4    h  1600000.0     VB   
...      ...            ...               ...    ...  ...        ...    ...   
13575  13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG    

Получение описательной статистики

→ Часто при работе с таблицей нужно быстро посмотреть на основные статистические свойства её столбцов. Для этого можно воспользоваться методом DataFrame describe().

По умолчанию метод работает с числовыми (int64 и float64) столбцами и показывает число непустых значений (count), среднее (mean), стандартное отклонение (std), минимальное значение (min),  квартили уровней 0.25, 0.5 (медиана) и 0.75 (25%, 50%, 75%) и максимальное значение (max) для каждого столбца исходной таблицы.

Чтобы не увязнуть в обилии информации, выведем на экран значение статистических параметров только для столбцов Distance (расстояние от объекта недвижимости до центра Мельбурна), BuildingArea (площадь здания) и Price (цена объекта):

In [9]:
print(melb_data.loc[:, ['Distance', 'BuildingArea', 'Price']])

       Distance  BuildingArea      Price
0           2.5         126.0  1480000.0
1           2.5          79.0  1035000.0
2           2.5         150.0  1465000.0
3           2.5         126.0   850000.0
4           2.5         142.0  1600000.0
...         ...           ...        ...
13575      16.7         126.0  1245000.0
13576       6.8         133.0  1031000.0
13577       6.8         126.0  1170000.0
13578       6.8         157.0  2500000.0
13579       6.3         112.0  1285000.0

[13580 rows x 3 columns]


In [11]:
print(melb_data.describe(include = ['object']))

           Suburb           Address   Type Method SellerG        Date  \
count       13580             13580  13580  13580   13580       13580   
unique        314             13378      3      5     268          58   
top     Reservoir  36 Aberfeldie St      h      S  Nelson  27/05/2017   
freq          359                 3   9449   9022    1565         473   

       CouncilArea             Regionname         Coordinates  
count        12211                  13580               13580  
unique          33                      8               13097  
top       Moreland  Southern Metropolitan  -37.8361, 144.9966  
freq          1163                   4695                  12  


Получение частоты уникальных значений в столбце

→ Для того чтобы определить, сколько раз в столбце повторяется каждый из вариантов значений (т.е. найти частоту для каждого уникального знания), используется метод value_counts().
Данный метод возвращает объект Series, в котором в качестве индексов выступают уникальные категории столбца, а значениями — соответствующая им частота.

Рассмотрим работу value_counts() на примере столбца с названиями районов:

In [15]:
print(melb_data['Regionname'].value_counts)

<bound method IndexOpsMixin.value_counts of 0             Northern Metropolitan
1             Northern Metropolitan
2             Northern Metropolitan
3             Northern Metropolitan
4             Northern Metropolitan
                    ...            
13575    South-Eastern Metropolitan
13576          Western Metropolitan
13577          Western Metropolitan
13578          Western Metropolitan
13579          Western Metropolitan
Name: Regionname, Length: 13580, dtype: object>



Чтобы сделать вывод более интерпретируемым и понятным, можно воспользоваться параметром normalize. При установке значения этого параметра на True результат будет представляться в виде доли (относительной частоты):

In [16]:
print(melb_data['Regionname'].value_counts(normalize= True))

Southern Metropolitan         0.345729
Northern Metropolitan         0.286451
Western Metropolitan          0.217084
Eastern Metropolitan          0.108321
South-Eastern Metropolitan    0.033137
Eastern Victoria              0.003903
Northern Victoria             0.003019
Western Victoria              0.002356
Name: Regionname, dtype: float64


МОДУЛЬ 7


Агрегирующим в Pandas называется метод, который для каждого столбца возвращает только одно значение — показатель (например, вычисление медианы, максимума, среднего и так далее).

Ниже приведена таблица основных агрегирующих методов:
Метод 	Статистический параметр
.count() 	Количество непустых значений
.mean() 	Среднее значение
.min() 	Минимальное значение
.max() 	Максимальное значение
.deviance() 	Дисперсия
.std() 	Стандартное отклонение
.sum() 	Сумма
.quantile(x) 	Квантиль уровня x
.nunique() 	Число уникальных значений

Если один из этих методов применить ко всему DataFrame, то в результате его работы будет получен объект типа Series, в котором в качестве индексов будут выступать наименования столбцов, а в качестве значений — статистический показатель. В случае применения метода к отдельному столбцу результатом вычислений станет число.

В каждый метод можно передать некоторые параметры, среди которых:

        axis  — определяет, подсчитывать параметр по строкам или по столбцам;
        numeric_only — определяет, вычислять параметры только по числовым столбцам/строкам или нет (True/False).


азберём агрегирующие функции на примерах.

Вычислим среднюю цену на объекты недвижимости:

In [17]:
print(melb_data['Price'].mean())
#Найдём максимальное количество парковочных мест:
print(melb_data['Car'].max())
#А теперь представим, что риэлторская ставка для всех компаний за продажу недвижимости составляет 12%. Найдём общую прибыльность риэлторского бизнеса в Мельбурне. Результат округлим до сотых:
rate = 0.12
income = melb_data['Price'].sum() * rate
print('Total income of real estate agencies:', round(income, 2))
landsize_median = melb_data['Landsize'].median() 
landsize_mean =  melb_data['Landsize'].mean()
print(abs(landsize_median - landsize_mean)/landsize_mean)

1075684.079455081
10.0
Total income of real estate agencies: 1752934775.88
0.21205713983546193


In [18]:
print(melb_data['Rooms'].mode())

0    3
dtype: int64


Модуль 8

Часто возникает необходимость исследовать определённую группу объектов по какому-то условию, например найти здания с ценой меньше 1 миллиона или выделить из всей таблицы помещения с двумя комнатами.

Такие задачи называются задачами фильтрации.

Под фильтрацией в DataFrame подразумевается получение новой таблицы путём вырезания строк, не удовлетворяющих поставленному условию. 

Разберём классический способ фильтрации в DataFrame — фильтрацию с помощью масок.

Маской называется Series, которая состоит из булевых значений, при этом значения True соответствуют тем индексам, для которых заданное условие выполняется, в противном случае ставится значение False (например, цена > 2 млн).

Создадим маску и положим её в переменную с именем mask. Синтаксис очень прост:

In [20]:
mask = melb_data['Price'] > 2000000
print(melb_data[mask].head)

<bound method NDFrame.head of        index          Suburb               Address  Rooms Type      Price  \
80        80     Albert Park  112 Beaconsfield Pde      3    h  2850000.0   
85        85     Albert Park     104 Richardson St      4    h  2300000.0   
88        88     Albert Park        29 Faussett St      2    h  2120000.0   
92        92     Albert Park           2 Dundas Pl      3    h  2615000.0   
93        93     Albert Park          23 Finlay St      5    h  2100000.0   
...      ...             ...                   ...    ...  ...        ...   
13521  13521  Port Melbourne          44 Garton St      4    t  2455000.0   
13523  13523         Prahran        69 Greville St      4    h  2668000.0   
13553  13553    Surrey Hills          20 Albert Cr      4    h  2720000.0   
13555  13555    Surrey Hills              3 Oak St      4    h  3100000.0   
13578  13578    Williamstown          96 Verdon St      4    h  2500000.0   

      Method   SellerG        Date  Distance 

In [7]:
print(melb_data[melb_data['Rooms'] == 3].shape[0])
#Усложним прошлый пример и найдём число трёхкомнатных домов с ценой менее 300 тысяч:
m = melb_data[(melb_data['Rooms'] == 3) & (melb_data['Price'] < 300000)].shape[0]
print(m)
m1 = melb_data[((melb_data['Rooms'] == 3) | (melb_data['BuildingArea'] > 100)) & (melb_data['Price'] < 300000)].shape[0]
print(m1)

print(melb_data[melb_data['Type'] == 't']['Rooms'].max())

mean_price = melb_data['Price'].mean()
melb_data[melb_data['Price'] > mean_price]['BuildingArea'].median()


5881
3
68
5


126.0